### PR TITLE
XWIKI-20282: Provide an API allowing to attach an XWikiAttachment as temporary attachment

### PR DESCRIPTION
**Pull request by surli**
Example usage of this new API: https://github.com/xwiki-contrib/application-changerequest/pull/74/commits/0a116bc826dd61649d0a429a0816672466b0f15f

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java [hunk 1]
```diff
@@ -74,6 +74,22 @@ XWikiAttachment uploadAttachment(DocumentReference documentReference, Part part)
     XWikiAttachment uploadAttachment(DocumentReference documentReference, Part part, String filename)
         throws TemporaryAttachmentException;
 
+    /**
+     * Allow to temporarily attach the given instance of {@link XWikiAttachment} to the given document reference.
+     * This can be useful if an API manipulates an {@link XWikiAttachment} without saving it, and want it to be
+     * discoverable when parsing a document through the download action for example.
+     * Note that consumer of this API needs to be aware that the file attached to the {@link XWikiAttachment} might be
+     * deleted at the end of the session, as any temporary attachment.
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java:82]: Is there a risk that the temporary attachments are viewable by other means (until the session is ended).
For instance, the attachment picker is now listing temporary attachments, could they be displayed there?

File: xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java [hunk 2]
```diff
@@ -74,6 +74,22 @@ XWikiAttachment uploadAttachment(DocumentReference documentReference, Part part)
...
+     *
+     * @param attachment the attachment to be temporarily attached to the document
+     * @param documentReference the reference of the document to link this attachment to
+     * @throws TemporaryAttachmentException in case of problem when performing the link
+     * @since 14.10RC1
+     */
+    @Unstable
+    void temporarilyAttach(XWikiAttachment attachment, DocumentReference documentReference)
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java:90]: probably missing a `revapi.differences` item?

File: xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java (hunk 1 above)
Comment (surli) [reply xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java:82]: Yes of course, the logic of the temporary attachment session doesn't change, so it's responsibility of the consumer of the API to be careful about what he's doing and to properly call the API to delete the temporary attachment when needed.

File: xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java (hunk 2 above)
Comment (surli) [reply xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java:90]: Right, I forgot to run legacy-oldcore. Thanks for the heads up

File: xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java (hunk 1 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/TemporaryAttachmentSessionsManager.java:82]: thanks, sounds good

File: xwiki-platform-core/xwiki-platform-store/xwiki-platform-store-filesystem-oldcore/src/test/java/org/xwiki/store/filesystem/internal/DefaultTemporaryAttachmentSessionsManagerTest.java [hunk 3]
```diff
@@ -254,4 +256,20 @@ void removeUploadedAttachments()
         when(temporaryAttachmentSession.removeAttachments(documentReference)).thenReturn(true);
         assertTrue(this.attachmentManager.removeUploadedAttachments(documentReference));
     }
+
+    @Test
+    void temporarilyAttach() throws TemporaryAttachmentException
+    {
+        DocumentReference documentReference = mock(DocumentReference.class);
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-store/xwiki-platform-store-filesystem-oldcore/src/test/java/org/xwiki/store/filesystem/internal/DefaultTemporaryAttachmentSessionsManagerTest.java:263]: Why a mock when an actual object could be easily instantiated?

**Review by manuelleduc — APPROVED** [review]

File: xwiki-platform-core/xwiki-platform-store/xwiki-platform-store-filesystem-oldcore/src/test/java/org/xwiki/store/filesystem/internal/DefaultTemporaryAttachmentSessionsManagerTest.java (hunk 3 above)
Comment (surli) [reply xwiki-platform-core/xwiki-platform-store/xwiki-platform-store-filesystem-oldcore/src/test/java/org/xwiki/store/filesystem/internal/DefaultTemporaryAttachmentSessionsManagerTest.java:263]: I always use mocks when I don't actually need an actual object in tests. That's how I see unit tests: anything not needed should be mocked as most as possible.

**Pull request merged**